compiler/compiler.go

```diff
@@ -0,0 +1,213 @@
+// Package compiler creates FileDescriptorProtos from a parser.Parse()
+// result. It is assumed that the parse result is semantically
+// correct and would compile with protoc.
+package compiler
+
+import (
+	"strings"
+
+	"github.com/alecthomas/protobuf/parser"
+	pb "google.golang.org/protobuf/types/descriptorpb"
+)
+
+type fdBuilder struct {
+	proto3   bool
+	fileDesc *pb.FileDescriptorProto
+}
+
+type messageBuilder struct {
+	proto3      bool
+	fullName    string
+	messageDesc *pb.DescriptorProto
+}
+
+func NewFileDescriptor(file string, p *parser.Proto) *pb.FileDescriptorProto {
+	b := &fdBuilder{
+		proto3: p.Syntax == "proto3",
+	}
+	b.buildFileDescriptor(file, p)
+	return b.fileDesc
+}
+
+func (b *fdBuilder) buildFileDescriptor(file string, p *parser.Proto) {
+	b.fileDesc = newFileDescritpor(file, b.proto3)
+	for _, e := range p.Entries {
+		b.addEntry(e)
+	}
+}
+
+func newFileDescritpor(file string, proto3 bool) *pb.FileDescriptorProto {
```

Comment:
nit: typo

Also, it's not clear to me why this isn't just in the method above?

Reply:
yeah, fair, too much re-jigging. fixed.